source/lac/lapack_full_matrix.cc

```diff
@@ -623,11 +623,66 @@ template <typename number>
 void
 LAPACKFullMatrix<number>::Tmmult(LAPACKFullMatrix<number>       &C,
                                  const LAPACKFullMatrix<number> &B,
+                                 const Vector<number>           &V,
                                  const bool                      adding) const
 {
   Assert(state == matrix || state == inverse_matrix, ExcState(state));
-  Assert(B.state == matrix || B.state == inverse_matrix, ExcState(state));
-  Assert(C.state == matrix || C.state == inverse_matrix, ExcState(state));
+  Assert(B.state == matrix || B.state == inverse_matrix, ExcState(B.state));
+  Assert(C.state == matrix || C.state == inverse_matrix, ExcState(C.state));
+
+  const LAPACKFullMatrix<number> &A = *this;
+
+  Assert (A.m() == B.m(), ExcDimensionMismatch(A.m(), B.m()));
+  Assert (C.n() == B.n(), ExcDimensionMismatch(C.n(), B.n()));
+  Assert (C.m() == A.n(), ExcDimensionMismatch(A.n(), C.m()));
+  Assert (V.size() == A.m(), ExcDimensionMismatch(A.m(), V.size()));
+  const unsigned int mm = A.n();
+  const unsigned int nn = B.n();
+  const unsigned int kk = B.m();
+
+  // lapack does not have any tripple product routines, including the case of
+  // diagonal matrix in the middle, see
+  // https://stackoverflow.com/questions/3548069/multiplying-three-matrices-in-blas-with-the-middle-one-being-diagonal
+  // http://mathforum.org/kb/message.jspa?messageID=3546564
+  // below just do a tripple loop.
+  if (!adding)
+    C = number(0.);
+  if (PointerComparison::equal (&A, &B))
+    {
+      Assert(mm==nn, ExcInternalError());
+      // symmetric case (A==B):
+      // Cij = Aki Vk Akj
+      for (unsigned int i = 0; i < mm; ++i)
+        for (unsigned int j = i; j < nn; ++j)
+          for (unsigned int k = 0; k < kk; ++k)
+            C(i,j) += A(k,i) * V(k) * A(k,j);
+
+      for (unsigned int i = 0; i < mm; ++i)
+        for (unsigned int j = i+1; j < nn; ++j)
+          C(j,i) = C(i,j);
+    }
```

Comment:
I don't really like to use a hand-written mat-mat-mult code in the context of LAPACK matrices because it is surely a factor 10, maybe even 20-40, slower than what a proper BLAS will do (at least for moderate to large matrix sizes). How about doing what the Stackoverflow suggests - which is also what I would have done - and simply copy the matrix B, multiply it by `V` (in n^2 operations) and then go to BLAS, i.e., the classical `Tmmult` multiplication, for the expensive n^3 operations? Of course, the copy will be visible for small sizes, but pretty soon the better implementation should take over, I guess once `m*n*k` is larger than 1000 (i.e, 10x10 matrices).

Reply:
> copy the matrix B, multiply it by V (in n^2 operations) and then go to BLAS, i.e., the classical Tmmult multiplication, for the expensive n^3 operations

i will update.